File: src/LeetCode/Algorithms/Java/Min_Stack.java
package LeetCode.Algorithms.Java;

import java.util.LinkedList;

// https://leetcode.com/problems/min-stack/
public class Min_Stack {
	public static void main(String[] args) {
		MinStack minStack = new MinStack();
		
		minStack.push(-2);
		minStack.push(0);
		minStack.push(-3);
		System.out.println(minStack.getMin());
		minStack.pop();
		System.out.println(minStack.top());
		System.out.println(minStack.getMin());
	}

	static class MinStack {
		private long min;
		LinkedList<Long> stack;
		
		public MinStack() {
			stack = new LinkedList<>();
		}
		
		public void push(int x) {
			if (stack.isEmpty()) {
				stack.push(0L);
				this.min = x;
			} else {
				stack.push(x - this.min);
				
				if (x < this.min) {
					this.min = x;
				}
			}
		}
		
		public void pop() {
			if (stack.isEmpty()) {
				return;
			}
			
			long pop = stack.pop();
			
			if (pop < 0) {
				this.min -= pop;
			}
		}
		
		public int top() {
			long top = stack.peek();
			
			if (top > 0) {
				return (int) (top + this.min);
			} else {
				return (int) this.min;
			}
		}
		
		public int getMin() {
			return (int) this.min;
		}
	}
}

File: src/LeetCode/Algorithms/JavaScript/Min_Stack.js
/**
 * https://leetcode.com/problems/min-stack/
 * initialize your data structure here.
 */
var MinStack = function() {
	this.min = Infinity;
	this.stack = new Array();
};

/**
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
	this.min = Math.min(this.min, x);
	this.stack.push({
		val: x
		, min : this.min
	});
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
	this.stack.pop();
	
	if (this.stack.length > 0) {
		this.min = this.stack[this.stack.length - 1].min;
	} else {
		this.min = Infinity;
	}
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
	return this.stack[this.stack.length - 1].val;
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
	return this.min;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */